### ISSUE TITLE
How to run sandstorm in the foreground?

### ISSUE THREAD

**Issue by abliss** [sandstorm-io/sandstorm#2839]
`sandstorm/sandstorm start` automatically forks itself into the background, but I'd like a way to keep it in the foreground so I can manage it with e.g. supervisord. I looked through the docs but couldn't find such a configuration option.

**Comment by abliss** [comment]
A local bash wiz helped me figure out this cool workaround:
`(mkfifo sandstorm-pipe; sandstorm/sandstorm start {foo}>sandstorm-pipe & cat sandstorm-pipe)`

But I think this only works because when `sandstorm` daemonizes itself, it doesn't close all open file descriptors, which seems like a bug?

**Comment by kentonv** [comment]
I'd accept a PR to add a `--foreground`.

@mrdomino I think this was something you were interested in, too? Or am I misremembering?

**Comment by mrdomino** [comment]
Yes, I'm interested. It'd make it easier to run Sandstorm as a [runit](http://smarden.org/runit/) service.

Hi @abliss!

Note that adding a `--foreground` is not trivial since sandstorm depends on being PID 1 in its own namespace. I gave it a crack once but didn't finish. I estimate it'd take a day or two's work, possibly less for someone more familiar with sandstorm and/or UNIX internals.

**Comment by kentonv** [comment]
Ah, right... we still need to fork in order to create a PID namespace where the Sandstorm process is PID 1 (which in turn means that killing the Sandstorm process also kills all its children, a desirable trait).

So for `--foreground`, I suppose what we'd want is to have the parent process `wait()` on the child rather than exiting after fork. (And don't redirect stdout/stderr.)

**Comment by kentonv** [comment]
Though I suppose you'd then want some additional mechanism that forces the child process to quit if the parent dies. I think there's a `prctl()` for that, maybe.

**Comment by abliss** [comment]
I just found out about https://github.com/catern/supervise which looks really cool and might be helpful in this context.

**Comment by abliss** [comment]
(How does sandstorm create a PID namespace without CAP_SYS_ADMIN ? I thought that was required.)

**Comment by zarvox** [comment]
> (How does sandstorm create a PID namespace without CAP_SYS_ADMIN ? I thought that was required.)

Unprivileged user namespaces, inside of which you have CAP_SYS_ADMIN.  Otherwise, you have to initially launch Sandstorm as root.
